### DISCUSSION TITLE
How to get graphs in distance mode and delta

### DISCUSSION THREAD

**Discussion by MaxZec** [theOehrly/Fast-F1#89] [Q&A]
Hi there,

I was wondering if it's possibile to draw graphs using data from real F1 in distance mode rather than time mode...this would make laps comparison much more accurate, and also it would give us the chance to calculate a proper delta which is impossibile in time mode...like in the image attached here, it would looks like that one is braking\accelerating later than the other, but we can't be sure coz the graph and data are in time mode and not distance mode....and is there any chance to have some data where corners are so we can easily add T1 T2 etc rather than do it manually? Thank you in advance!
![Senza titolo-3](https://user-images.githubusercontent.com/95567617/144742352-fb0fdd68-6156-4f53-8676-22fc3b69ca97.jpg)

**Comment by theOehrly** [comment]
Yes, you can plot data over distance. I thought there was an example for that, but I can't find it. So I guess there's actually no example for it.

The recommended way is this
```
lap = laps.pick_driver('VER').pick_fastest()
tel = lap.get_car_data().add_distance()
plt.plot(tel['Distance'], tel['Speed'])
```

Parts of this article in medium.com might be interesting:
https://medium.com/towards-formula-1-analysis/how-i-analyze-formula-1-data-with-python-2021-italian-gp-dfb11db4b73


There is no data on corner numbers. It's currently not possible to add automated labels for that.

**Comment by MaxZec** [comment]
Thank you for your reply! I've done a test but I believe it's not useful for what I meant. As the data record starts in different points (it seems), the X axis (that should be the same for everybody) won't match...see the image...first data recorded for VER lap is different from HAM's one...so I can't use distance data to have a distance mode or calculate a time delta...
![Immagine](https://user-images.githubusercontent.com/95567617/144748417-be4e7d8a-71ce-4221-8c31-6120f7cdc0d4.jpg)
..

**Comment by theOehrly** [comment]
I don't quite understand what you want. Do you want the distance to start a zero and then increment in constant intervals?

**Comment by MaxZec** [comment]
Something like that yes, unless I'm missing something that can make me use that difference distances values. I'll try to make myself clear sorry.

So, if I have to compare let's say how and where HAM and VER  brake in corner 1, especially in terms of meters before the corner...to do this I need that all the data (brakes, throttle, rpm, gears etc) started from the same point and being constant through all the record data.. So let's say from Mt 5 after the start line and increment of 2 meters every record data.. So that I'm sure that if HAM brakes at 50mt and and VER a bit later means he brakes earlier...in time mode it's impossible to know this...

**Comment by theOehrly** [comment]
What you want to do is possible in theory, of course. But FastF1 does not provide any means for you to do it. The reasons for this is the low sampling rate and comparably low accuracy of the data (look at the speed traces, they are jittery).
The distance is calculated by integrating speed over time. The result is good enough for plotting but not good enough to base any further calculations on. Especially not if they require high accuracy. Determining braking points requires very high accuracy.
If you were to resample the distance values into constant intervals, the result would be highly inaccurate. You are starting with data that is not perfectly accurate, and then you introduce more inaccuracy through resampling.

Furthermore, let me give you a short calculation as an example for why the data is not accurate enough to determine braking points. 
The source data has a sample rate of 5 Hz (actually slightly lower even). That means, we have one sample every 200 milliseconds.
Assume that the car is driving at 300 km/h. That is equivalent to 83 m/s. Now, let's assume, that your data is shifted by only one sample or 1/5th of a second. 83 m/s times 0.2 seconds is 17 m. That is a huge difference!
The data is not perfectly synchronized. So you will have some inaccuracies there. The comparison that you want to do is not possible with the data we have. We'd need some well synchronized data at a considerably higher sampling rate, but the api does not provide that.

My current assumption is, that the accuracy is about +-0.05 seconds (there are outliers).

**Comment by MaxZec** [comment]
Thank you very much for clarify!!!

**Comment by OmaMansou** [comment]
Hello Max, 
I think I know what you want to do, cause I tried doing the same, but it doesn't seem like there's a solution that gives 100% accuracy.

One method I found (and that gave satisfactory results) was merging the tables on timings (for driver_1 and driver_2) based on what time is closest using pandas.merge_asof and then applying a smoothing. Here is how i did it : 

    times_1 = telemetry_1[['Time', 'Distance']].rename(columns={'Time': 'time_1'})

    times_2 = telemetry_2[['Time', 'Distance']].rename(columns={'Time': 'time_2'})

    times = merge_asof(left=times_1,
                                     right=times_2,
                                     on='Distance',
                                     direction='nearest')

    times.loc[:, 'delta'] = times.time_2 - times.time_1

    times.loc[:, 'smooth_delta'] = times.delta.rolling(20, center=True).mean()

The delta column is very jittery, but once the smoothing is applied (with rolling and mean), the plot looks just fine and is quite usable.

**Comment by OmaMansou** [comment]
I managed to get this plot for the qualifying session of the 2021 Bahrain GP.

![image](https://user-images.githubusercontent.com/42304770/148252016-cd84cee1-fbd5-47dc-981c-939a002389eb.png)

**Reply by MaxZec** [reply]
> Reference in

Hello, a friend of mine who is trying to help me got this error...

Traceback (most recent call last): 
  File "C:\Python39\lib\site-packages\pandas\core\window\rolling.py", line 427, in _apply_series 
    values = self._prep_values(obj._values) 
  File "C:\Python39\lib\site-packages\pandas\core\window\rolling.py", line 337, in _prep_values 
    raise NotImplementedError( 
NotImplementedError: ops for Rolling for this dtype timedelta64[ns] are not implemented 
 
The above exception was the direct cause of the following exception: 
 
Traceback (most recent call last): 
  File "C:\Shared\src\QA\F1Drivers.py", line 181, in <module> 
    main() 
  File "C:\Shared\src\QA\F1Drivers.py", line 177, in main 
    extract_and_generate(args) 
  File "C:\Shared\src\QA\F1Drivers.py", line 91, in extract_and_generate 
    times.loc[:, 'smooth_delta'] = times.delta.rolling(20, center=True).mean() 
  File "C:\Python39\lib\site-packages\pandas\core\window\rolling.py", line 2039, in mean 
    return super().mean(*args, engine=engine, engine_kwargs=engine_kwargs, **kwargs) 
  File "C:\Python39\lib\site-packages\pandas\core\window\rolling.py", line 1450, in mean 
    return self._apply(window_func, name="mean", **kwargs) 
  File "C:\Python39\lib\site-packages\pandas\core\window\rolling.py", line 590, in _apply 
    return self._apply_blockwise(homogeneous_func, name) 
  File "C:\Python39\lib\site-packages\pandas\core\window\rolling.py", line 442, in _apply_blockwise 
    return self._apply_series(homogeneous_func, name) 
  File "C:\Python39\lib\site-packages\pandas\core\window\rolling.py", line 429, in _apply_series 
    raise DataError("No numeric types to aggregate") from err 
pandas.core.base.DataError: No numeric types to aggregate

the problem probably lie in this line

times.loc[:, 'smooth_delta'] = times.delta.rolling(20, center=True).mean()

if he "comment" the line it works but delta filed  gives very strange values....any help??

**Reply by ezkomoly** [reply]
> > Reference in
> 
> Hello, a friend of mine who is trying to help me got this error...
> 
> Traceback (most recent call last): File "C:\Python39\lib\site-packages\pandas\core\window\rolling.py", line 427, in _apply_series values = self._prep_values(obj._values) File "C:\Python39\lib\site-packages\pandas\core\window\rolling.py", line 337, in _prep_values raise NotImplementedError( NotImplementedError: ops for Rolling for this dtype timedelta64[ns] are not implemented
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last): File "C:\Shared\src\QA\F1Drivers.py", line 181, in main() File "C:\Shared\src\QA\F1Drivers.py", line 177, in main extract_and_generate(args) File "C:\Shared\src\QA\F1Drivers.py", line 91, in extract_and_generate times.loc[:, 'smooth_delta'] = times.delta.rolling(20, center=True).mean() File "C:\Python39\lib\site-packages\pandas\core\window\rolling.py", line 2039, in mean return super().mean(*args, engine=engine, engine_kwargs=engine_kwargs, **kwargs) File "C:\Python39\lib\site-packages\pandas\core\window\rolling.py", line 1450, in mean return self._apply(window_func, name="mean", **kwargs) File "C:\Python39\lib\site-packages\pandas\core\window\rolling.py", line 590, in _apply return self._apply_blockwise(homogeneous_func, name) File "C:\Python39\lib\site-packages\pandas\core\window\rolling.py", line 442, in _apply_blockwise return self._apply_series(homogeneous_func, name) File "C:\Python39\lib\site-packages\pandas\core\window\rolling.py", line 429, in _apply_series raise DataError("No numeric types to aggregate") from err pandas.core.base.DataError: No numeric types to aggregate
> 
> the problem probably lie in this line
> 
> times.loc[:, 'smooth_delta'] = times.delta.rolling(20, center=True).mean()
> 
> if he "comment" the line it works but delta filed gives very strange values....any help??

Hi Max! 
I'm new to python, so I'm not 100% thats the issue, but I had the same error with the code above.
I managed to resolve it, by converting the 'delta' format from 0 days 00:00:00.000000 to seconds.

**Reply by OmaMansou** [reply]
Oh sorry, I failed to point that out in my comment. I converted the times to milliseconds, so the delta is actually in milliseconds. But converting it to seconds would also work indeed.

Suppose the name of the column in a dataframe df is called Time, I converted the value to milliseconds by :

`df.loc[:, 'Time'] = df.Time.dt.seconds * 1000 + df.Time.dt.microseconds / 1000`

Tell me if this solves the bug in your case as well.

**Reply by MaxZec** [reply]
Thank you both, I'll pass the info and see if it works!

**Reply by MaxZec** [reply]
Hi there! thnx for the info, but we still facing some troubles with the smooth delta...here's what we got:

TIMES: 
       time_1     Distance    time_2   delta  smooth_delta 
0       99.0     0.000000     299.0   200.0           NaN 
1      299.0     0.000000     299.0     0.0           NaN 
2      739.0     1.222222     739.0     0.0           NaN 
3     1059.0     3.177778    1059.0     0.0           NaN 
4     1379.0     6.022222    1379.0     0.0           NaN 
..       ...          ...       ...     ...           ... 
362  98019.0  5043.958056   99219.0  1200.0           NaN 
363  98259.0  5061.891389   99459.0  1200.0           NaN 
364  98539.0  5082.891389   99779.0  1240.0           NaN 
365  98979.0  5116.380278  100139.0  1160.0           NaN 
366  99219.0  5134.646944  100459.0  1240.0           NaN

this is what we have used:

times1 = telemetry1[['Time', 'Distance']].rename(columns={'Time': 'time_1'})
times1.loc[:, 'time_1'] = times1.time_1.dt.seconds * 1000 + times1.time_1.dt.microseconds / 1000
times2 = telemetry2[['Time', 'Distance']].rename(columns={'Time': 'time_2'})
times2.loc[:, 'time_2'] = times2.time_2.dt.seconds * 1000 + times2.time_2.dt.microseconds / 1000
times = pd.merge_asof(left=times1, right=times2, on='Distance', direction='nearest')
times.loc[:, 'delta'] = times.time_2 - times.time_1
times.loc[:, 'smooth_delta'] = times.delta.rolling(20, center=True).mean()


for row, delta_row in zip(telemetry.itertuples(), times.itertuples()):
    values = list(row)[1:-3] + [row[-1], delta_row[-1]]

**Reply by MaxZec** [reply]
> I managed to get this plot for the qualifying session of the 2021 Bahrain GP.
> 
> ![image](https://user-images.githubusercontent.com/42304770/148252016-cd84cee1-fbd5-47dc-981c-939a002389eb.png)

Another question...how have you been able to match the data? I mean looks like a proper to distance graph and not to time graph...if you check my preview screen I always have lot of differences in where drivers apply input, like if was a different track at all...while your graphs are damn close! :)

**Reply by OmaMansou** [reply]
> Hi there! thnx for the info, but we still facing some troubles with the smooth delta...here's what we got:
> 
> TIMES: time_1 Distance time_2 delta smooth_delta 0 99.0 0.000000 299.0 200.0 NaN 1 299.0 0.000000 299.0 0.0 NaN 2 739.0 1.222222 739.0 0.0 NaN 3 1059.0 3.177778 1059.0 0.0 NaN 4 1379.0 6.022222 1379.0 0.0 NaN .. ... ... ... ... ... 362 98019.0 5043.958056 99219.0 1200.0 NaN 363 98259.0 5061.891389 99459.0 1200.0 NaN 364 98539.0 5082.891389 99779.0 1240.0 NaN 365 98979.0 5116.380278 100139.0 1160.0 NaN 366 99219.0 5134.646944 100459.0 1240.0 NaN
> 
> this is what we have used:
> 
> times1 = telemetry1[['Time', 'Distance']].rename(columns={'Time': 'time_1'}) times1.loc[:, 'time_1'] = times1.time_1.dt.seconds * 1000 + times1.time_1.dt.microseconds / 1000 times2 = telemetry2[['Time', 'Distance']].rename(columns={'Time': 'time_2'}) times2.loc[:, 'time_2'] = times2.time_2.dt.seconds * 1000 + times2.time_2.dt.microseconds / 1000 times = pd.merge_asof(left=times1, right=times2, on='Distance', direction='nearest') times.loc[:, 'delta'] = times.time_2 - times.time_1 times.loc[:, 'smooth_delta'] = times.delta.rolling(20, center=True).mean()
> 
> for row, delta_row in zip(telemetry.itertuples(), times.itertuples()): values = list(row)[1:-3] + [row[-1], delta_row[-1]]

This is actually working! When you're applying rolling, it calculates the mean value over a window of (20 in this case) points, and in this case it is centered (10 on each side). So for the first and last 10 points, it won't be able to calculate the delta. But if you try to see the content of the other lines, I think it'll have valid data.

**Reply by OmaMansou** [reply]
> > I managed to get this plot for the qualifying session of the 2021 Bahrain GP.
> > ![image](https://user-images.githubusercontent.com/42304770/148252016-cd84cee1-fbd5-47dc-981c-939a002389eb.png)
> 
> Another question...how have you been able to match the data? I mean looks like a proper to distance graph and not to time graph...if you check my preview screen I always have lot of differences in where drivers apply input, like if was a different track at all...while your graphs are damn close! :)

All the plots are relative to distance. Time is only used when calculating the delta. Also, if you try to zoom in on my plots, they're not that perfect either.

If you'd like to see more plots with my code, I have personally developed a website that displays the telemetry : 
www.f1-telemetry.net/telemetry

**Reply by MaxZec** [reply]
Wow the web site is really cool, congrats for the wonderful job mate! Really impressive! When I use data to draw plots I got a lot of difference like this...

![image](https://user-images.githubusercontent.com/95567617/160280078-efc7396d-c91e-4c84-a235-956f66bcf1fa.png)

how can I get closer values like you?

**Reply by OmaMansou** [reply]
> Wow the web site is really cool, congrats for the wonderful job mate! Really impressive! When I use data to draw plots I got a lot of difference like this...
> 
> ![image](https://user-images.githubusercontent.com/95567617/160280078-efc7396d-c91e-4c84-a235-956f66bcf1fa.png)
> 
> how can I get closer values like you?

What type of data are you comparing ? Is it from practice sessions, qualifications, the race ? If you havent tried yet, try comparing the fastest laps for Verstappen and Leclerc in quali sessions, because they tend to be more similar.

**Reply by MaxZec** [reply]
![image](https://user-images.githubusercontent.com/95567617/160281763-ab58ffba-df5a-47a1-8fac-7019494e406f.png)

Yes, this is PER vs LEC yesterday. Data seems closer indeed. Looks like Checo did the magic in the last part braking earlier and carrying less speed at the apex but being able to accelerate earlier and carry more speed on the straight.

**Reply by markste-in** [reply]
Please don't use pandas merge_asof. It will just match to the nearest value and thus may create artefacts. A slightly better way would be to resample the signal and then interpolate. That will result in a smaller loss of accuracy.

**Comment by MaxZec** [comment]
Sorry again...trying to use the delta feature, and comparing Lap 18 from Perez (the pole lap) with the fastest of LEC (lap 20) i got some strange values. The final delta is of course just 25 mil...but this is what i got...never in the world they had such 0.400 of differernces in the lap, and also the data from delta stops much earlier than other datas (speed, throttle etc).

the label on the right and on top shows RPM but it's not

![image](https://user-images.githubusercontent.com/95567617/161025482-1206cabc-4961-4da3-9180-db8dd54de2b7.png)

**Reply by theOehrly** [reply]
You are comparing the fastest qualifying laps of Perez and Leclerc in the Saudi Arabia Grand Prix, did I understand that correctly? Please be precise about event, session, drivers and which laps you are looking at.
Also, a short snippet of code is always better than a verbal description only of what you are trying to do.

**Reply by MaxZec** [reply]
Yep sorry, Perez pole Vs Leclerc best lap. This is the code

times1 = telemetry1[['Time', 'Distance']].rename(columns={'Time': 'time_1'})
times1.loc[:, 'time_1'] = times1.time_1.dt.seconds * 1000 + times1.time_1.dt.microseconds / 1000
times2 = telemetry2[['Time', 'Distance']].rename(columns={'Time': 'time_2'})
times2.loc[:, 'time_2'] = times2.time_2.dt.seconds * 1000 + times2.time_2.dt.microseconds / 1000
times = pd.merge_asof(left=times1, right=times2, on='Distance', direction='nearest')
times.loc[:, 'delta'] = times.time_2 - times.time_1
times.loc[:, 'smooth_delta'] = times.delta.rolling(20, center=True).mean()
print('TIMES:\n', times)

for driver_symbol, lap_number, lap, telemetry in drivers_laps:
    for row, delta_row in zip(telemetry.itertuples(), times.itertuples()):
        values = list(row)[1:-3] + [row[-1], delta_row[-1]]

**Reply by markste-in** [reply]
What is the x-axis in your excel sheet? Looks like Sample-Nr or is it Time or Distance? If you plot it over sample nr I don't see anything wrong with it since the speed is quite different (and therefore also the distance over time)

**Reply by MaxZec** [reply]
> 

Hi, good point...what should i Insert? This is the data i got every time using the code above:

![image](https://user-images.githubusercontent.com/95567617/161387731-4992fbe4-b6fe-4ecd-98e0-8037929eea8d.png)

![image](https://user-images.githubusercontent.com/95567617/161387863-201c661d-acf3-493b-8093-2cb605243fd1.png)



so i don't really know what should the delta looks like...those numbers are quite weird to me, to be propere time delta should I insert the time in seconds or whaterver in the X axis?

those are the graph i'm abel to get...delta is driving me mad.

![image](https://user-images.githubusercontent.com/95567617/161388016-2f4e6528-59dc-40f2-852a-f3a7b720e775.png)

**Reply by MaxZec** [reply]
> What is the x-axis in your excel sheet? Looks like Sample-Nr or is it Time or Distance? If you plot it over sample nr I don't see anything wrong with it since the speed is quite different (and therefore also the distance over time)

Any help pls?

**Reply by markste-in** [reply]
Please post a minimal, but fully working code example that gives the same output as above. Sorry but I can't debug screenshots from an excel sheet.

**Reply by theOehrly** [reply]
@MaxZec sorry that I only get around to answering now. But I have to agree with @markste-in, a minimal but full code example to reproduce this is needed to understand what's going on. And we can't really help you with screenshots of excel sheets because we don't know what data is plotted where and how.

**Reply by MaxZec** [reply]
Thank you both, here's what you asked for (at least I hope so!)
[F1PoC2.zip](https://github.com/theOehrly/Fast-F1/files/8467886/F1PoC2.zip)

and those the columns and data it's able to create.

![image_2022-04-11_22-04-32](https://user-images.githubusercontent.com/95567617/162837059-498994a7-83aa-4026-b3cc-59b0cbf911a8.png)

**Reply by MaxZec** [reply]
@markste-in @theOehrly sorry to bother you guys, but is this what you have asked for?

**Reply by theOehrly** [reply]
@MaxZec I'll get back to you at some point. I just haven't had any time to look at this at all.

**Reply by markste-in** [reply]
I think I know what’s wrong (a few things) but I won’t have time during Easter. I’ll give u an answer next week

**Reply by siddharth-tewari** [reply]
@MaxZec Hello, would you find giving the code for how do you put data in excel like in this image? I need it for some physics project.
![image](https://user-images.githubusercontent.com/76590952/164885512-62df4576-1424-4cfc-80b1-ff3719603eee.png)

**Reply by MaxZec** [reply]
@theOehrly @markste-in sorry to be a pain the neck, just a reminder if you could point me at the error in the code....thnx!

**Reply by MaxZec** [reply]
> @MaxZec Hello, would you find giving the code for how do you put data in excel like in this image? I need it for some physics project. ![image](https://user-images.githubusercontent.com/76590952/164885512-62df4576-1424-4cfc-80b1-ff3719603eee.png)

Hi mate, I'll ask to the person who made it the permission to share!

**Reply by theOehrly** [reply]
@MaxZec You have to understand that this has very low priority for me. It's no efficient use of my time for this project. Debugging your scripts or excel sheets does not help anybody else who is using FastF1. So there is little incentive for me to invest time in it. I'm sorry, but I have really no idea when I'll get around to that.

**Reply by MaxZec** [reply]
I understand, thank you anyway. Just you asked for a full code that (I assume) I've posted above in a zip file, yet you mention script or excel sheets...thats why I was asking, that's all.

**Reply by MaxZec** [reply]
> I think I know what’s wrong (a few things) but I won’t have time during Easter. I’ll give u an answer next week

Hi there, any chance you could help to figure the error pls??

**Comment by MaxZec** [comment]
@markste-in could you pls help me with the error in delta data??
